ta/pkcs11/src/processing_digest.c

```diff
@@ -0,0 +1,250 @@
+#include <string.h>
+#include <tee_api_defines.h>
+#include <tee_internal_api.h>
+#include <tee_internal_api_extensions.h>
+#include <utee_defines.h>
+#include <util.h>
+
+#include "attributes.h"
+#include "object.h"
+#include "pkcs11_attributes.h"
+#include "pkcs11_helpers.h"
+#include "pkcs11_token.h"
+#include "processing.h"
+#include "serializer.h"
+
+bool processing_is_tee_digest(enum pkcs11_mechanism_id proc_id)
+{
+	switch (proc_id) {
+	case PKCS11_CKM_MD5:
+	case PKCS11_CKM_SHA_1:
+	case PKCS11_CKM_SHA224:
+	case PKCS11_CKM_SHA256:
+	case PKCS11_CKM_SHA384:
+	case PKCS11_CKM_SHA512:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static enum pkcs11_rc
+pkcs2tee_algorithm(uint32_t *tee_id, struct pkcs11_attribute_head *proc_params)
+{
+	static const struct {
+		enum pkcs11_mechanism_id mech_id;
+		uint32_t tee_id;
+	} pkcs2tee_algo[] = {
+		{ PKCS11_CKM_MD5, TEE_ALG_MD5 },
+		{ PKCS11_CKM_SHA_1, TEE_ALG_SHA1 },
+		{ PKCS11_CKM_SHA224, TEE_ALG_SHA224 },
+		{ PKCS11_CKM_SHA256, TEE_ALG_SHA256 },
+		{ PKCS11_CKM_SHA384, TEE_ALG_SHA384 },
+		{ PKCS11_CKM_SHA512, TEE_ALG_SHA512 },
+	};
+	size_t n = 0;
+
+	for (n = 0; n < ARRAY_SIZE(pkcs2tee_algo); n++) {
+		if (proc_params->id == pkcs2tee_algo[n].mech_id) {
+			*tee_id = pkcs2tee_algo[n].tee_id;
+			return PKCS11_CKR_OK;
+		}
+	}
+
+	return PKCS11_RV_NOT_IMPLEMENTED;
+}
+
+static enum pkcs11_rc
+allocate_tee_operation(struct pkcs11_session *session,
+		       struct pkcs11_attribute_head *params)
+{
+	uint32_t algo = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	assert(session->processing->tee_op_handle == TEE_HANDLE_NULL);
+
+	if (pkcs2tee_algorithm(&algo, params))
+		return PKCS11_CKR_FUNCTION_FAILED;
+
+	res = TEE_AllocateOperation(&session->processing->tee_op_handle,
+				    algo, TEE_MODE_DIGEST, 0);
+	if (res)
+		EMSG("TEE_AllocateOp. failed %#"PRIx32, algo);
+
+	if (res == TEE_ERROR_NOT_SUPPORTED)
+		return PKCS11_CKR_MECHANISM_INVALID;
+
+	return tee2pkcs_error(res);
+}
+
+enum pkcs11_rc init_digest_operation(struct pkcs11_session *session,
+				     enum processing_func function __unused,
+				     struct pkcs11_attribute_head *proc_params,
+				     struct pkcs11_object *ob __unused)
+{
+	assert(processing_is_tee_digest(proc_params->id));
+
+	return allocate_tee_operation(session, proc_params);
+}
+
+/*
+ * step_digest_operation - processing digest operation step
+ *
+ * @session - current session
+ * @step - step ID in the processing (oneshot, update, final)
+ * @ptype - invocation parameter types
+ * @params - invocation parameter references
+ */
+enum pkcs11_rc step_digest_operation(struct pkcs11_session *session,
+				     enum processing_func function __unused,
+				     enum processing_step step,
+				     struct pkcs11_object *obj,
+				     uint32_t ptypes, TEE_Param *params)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	void *in_buf = NULL;
+	size_t in_size = 0;
+	void *out_buf = NULL;
+	uint32_t out_size = 0;
```

Comment:
AFAICS this function is missing the handling described in section 5.2 of the [pkcs11 base spec](http://docs.oasis-open.org/pkcs11/pkcs11-base/v3.0/pkcs11-base-v3.0.html), namely that `out_size` should return a suitably large size if a `NULL` buffer is provided.

Reply:
That should be the digest size in this case. It kinda there but not there I suppose -- but there is easy way to verify that :)... I'll add the test case for this so we know for sure if it is functioning properly (and fix the problem).

Note: I believe we are still targeting 2.40. so not 3.0 yet but I believe this has not changed.

Reply:
AFAIR the current upstream symmetric `AES` implementation suffers from the same problem. The section is the same in 2.40, means I don't have to update my local copies just yet :grin:. Thanks for working on this :+1:

Reply:
Well... Added tests:

https://github.com/OP-TEE/optee_test/pull/491/files#diff-b95c9662712bacacfbfff42271a4da4d3aedf49989a6d7be645a69280096492eR4249

But it passes :)

The why it works is the side effect of calling `TEE_DigestDoFinal()`.

`When the function returns TEE_ERROR_SHORT_BUFFER, it SHALL return the size of the output data.`

Reply:
Ah, great, so the semantics match up :)

Another test case which uses `C_Digest`to query the digest size and we are golden. Thanks for adding the case.

Reply:
Added that extra test case.